lib/puppet/provider/ssh_authorized_key/parsed.rb

```diff
@@ -42,9 +42,27 @@ def file_perm
     0o600
   end
 
-  def user
-    uid = Puppet::FileSystem.stat(target).uid
-    Etc.getpwuid(uid).name
+  def group_writable_perm
+    0o020
+  end
+
+  def group_writable?(path)
+    path.stat.mode & group_writable_perm != 0
+  end
+
+  def trusted_path
+    return false unless Puppet::FileSystem.exist?(target)
+    path = Puppet::FileSystem.pathname(target).dirname
+    until path.dirname.root?
+      path = path.realpath if path.symlink?
+      # do not trust if path is world or group writable
+      if path.stat.uid != Process.euid || path.world_writable? || group_writable?(path)
+        Puppet.debug('Path not trusted. Will attempt to write as the target user.')
+        return false
+      end
+      path = path.dirname
+    end
+    Puppet.debug('Path trusted. Writing the file as root.')
```

Comment:
puppet usually runs as root, but not always. Maybe just say we're writing the file as the current user?